clients/drcachesim/tracer/syscall_pt_trace.cpp

```diff
@@ -0,0 +1,144 @@
+#define PT_METADATA_FILE_NAME_SUFFIX ".metadata"
+
+syscall_pt_trace_t::syscall_pt_trace_t()
+    : write_file_func_(nullptr)
+    , pttracer_handle_ {}
+    , recorded_syscall_num_(0)
+    , cur_recording_sysnum_(-1)
+    , drcontext_(nullptr)
+    , pt_dir_name_ { '\0' }
+{
+}
+
+syscall_pt_trace_t::~syscall_pt_trace_t()
+{
+}
+
+bool
+syscall_pt_trace_t::init(void *drcontext, char *pt_dir_name, size_t pt_dir_name_size,
+                         ssize_t (*write_file_func)(file_t file, const void *data,
+                                                    size_t count))
+{
+    drcontext_ = drcontext;
+    memcpy(pt_dir_name_, pt_dir_name, pt_dir_name_size);
+    write_file_func_ = write_file_func;
+    return true;
+}
+
+bool
+syscall_pt_trace_t::start_syscall_pt_trace(int sysnum)
+{
+#define RING_BUFFER_SIZE_SHIFT 8
+    ASSERT(drcontext_ != nullptr, "drcontext_ is nullptr");
+    ASSERT(pttracer_handle_.handle == nullptr, "pttracer_handle_.handle isn't nullptr");
+    if (drpttracer_create_handle(drcontext_, DRPTTRACER_TRACING_ONLY_KERNEL,
+                                 RING_BUFFER_SIZE_SHIFT, RING_BUFFER_SIZE_SHIFT,
+                                 &pttracer_handle_.handle) != DRPTTRACER_SUCCESS) {
+        return false;
+    }
+    if (drpttracer_start_tracing(drcontext_, pttracer_handle_.handle) !=
+        DRPTTRACER_SUCCESS) {
+        return false;
+    }
+    cur_recording_sysnum_ = sysnum;
+    return true;
+}
+
+bool
+syscall_pt_trace_t::stop_syscall_pt_trace()
+{
+    ASSERT(drcontext_ != nullptr, "drcontext_ is nullptr");
+    ASSERT(pttracer_handle_.handle != nullptr, "pttracer_handle_.handle is nullptr");
+
+    drpttracer_output_cleanup_last_t output;
+    if (drpttracer_stop_tracing(drcontext_, pttracer_handle_.handle, &output.data) !=
+        DRPTTRACER_SUCCESS) {
+        return false;
+    }
+    pttracer_handle_.~drpttracer_handle_cleanup_last_t();
+    pttracer_handle_ = {};
+    cur_recording_sysnum_ = -1;
+    recorded_syscall_num_++;
+    return trace_data_dump(output);
+}
+
+bool
+syscall_pt_trace_t::trace_data_dump(drpttracer_output_cleanup_last_t &output)
+{
+    ASSERT(drcontext_ != nullptr, "drcontext_ is nullptr");
+
+    drpttracer_output_t *data = output.data;
+
+    ASSERT(data->pt != nullptr, "pt is nullptr");
+    ASSERT(data->pt_size > 0, "pt_size is 0");
+    if (data->pt == nullptr || data->pt_size == 0) {
+        return false;
+    }
+
+    /* TODO i#5505: To reduce the overhead caused by IO, we better use a buffer to store
+     * multiple PT trace data. And only dump the buffer when it is full.
+     */
+
+    /* Dump PT trace data to file '{thread_id}.{syscall_id}.pt'. */
+    char pt_filename[MAXIMUM_PATH];
+    dr_snprintf(pt_filename, BUFFER_SIZE_ELEMENTS(pt_filename), "%s%s%d.%d%s",
+                pt_dir_name_, DIRSEP, dr_get_thread_id(drcontext_), recorded_syscall_num_,
+                PT_DATA_FILE_NAME_SUFFIX);
+    file_t pt_file = dr_open_file(pt_filename, DR_FILE_WRITE_REQUIRE_NEW);
```

Comment:
We can't use this here: all i/o including opening files must be redirected through the pointers registered with drmemtrace.

Reply:
Done.